ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/API.hs

```diff
@@ -274,61 +231,273 @@ iteratorToList it = go []
+-- in case of success or when an exception was raised.
+withDB ::
+     (HasCallStack, MonadThrow m)
+  => m (ImmutableDB m blk)
+     -- ^ How to open the database
+  -> (ImmutableDB m blk -> m a)
+     -- ^ Action to perform using the database
+  -> m a
+withDB openDB = bracket openDB closeDB
+
+getKnownBlock ::
+     (MonadThrow m, HasHeader blk)
+  => ImmutableDB m blk
+  -> RealPoint blk
+  -> m blk
+getKnownBlock db = getKnownBlockComponent db GetBlock
+
+getKnownHeader ::
+     (MonadThrow m, HasHeader blk)
+  => ImmutableDB m blk
+  -> RealPoint blk
+  -> m (Header blk)
+getKnownHeader db = getKnownBlockComponent db GetHeader
+
+getKnownBlockComponent ::
+     (MonadThrow m, HasHeader blk)
+  => ImmutableDB m blk
+  -> BlockComponent blk b
+  -> RealPoint blk
+  -> m b
+getKnownBlockComponent db blockComponent pt = do
+    mBlock <- mustExist pt <$> getBlockComponent db blockComponent pt
+    case mBlock of
+      Right b  -> return b
+      Left err -> throwM err
+
+-- | Open an iterator with the given point, which is known to be in the
+-- ImmutableDB, as lower exclusive bound and the current tip as the inclusive
+-- upper bound.
+--
+-- When the bound is invalid, a 'MissingBlockError' exception is thrown.
+streamAfterKnownBlock ::
+     (MonadSTM m, MonadThrow m, HasHeader blk, HasCallStack)
+  => ImmutableDB m blk
+  -> ResourceRegistry m
+  -> BlockComponent blk b
+  -> Point blk
+  -> m (Iterator m blk b)
+streamAfterKnownBlock db registry blockComponent fromPt = do
+    tipPt <- atomically $ getTipPoint db
+    case (pointToWithOriginRealPoint fromPt,
+          pointToWithOriginRealPoint tipPt) of
+
+      (Origin, Origin) ->
+        -- Nothing to stream, but no error either
+        return emptyIterator
+
+      (NotOrigin fromPt', Origin) ->
+        -- Asked to stream after a block while the ImmutableDB is empty, clearly
+        -- the block is missing.
+        throwUnexpectedError $ MissingBlockError fromPt'
+
+      (origFromPt, NotOrigin tipPt') -> do
+        -- Stream from the given point to the current tip (not genesis)
+        mItr <- stream
+                  db
+                  registry
+                  blockComponent
+                  (StreamFromExclusive fromPt)
+                  (StreamToInclusive tipPt')
+        case mItr of
+          -- If we got a 'WrongBoundError',
+          Left e
+            | NotOrigin fromPt' <- origFromPt
+              -- The block corresponding to the given point must not have been
+              -- in the ImmutableDB
+            -> throwUnexpectedError $ MissingBlockError fromPt'
+            | otherwise
+              -- If the point corresponds to genesis, then it's impossible to
+              -- get a WrongBoundError, as it must /always/ be possible to
+              -- stream from genesis to the current tip of the ImmutableDB.
+            -> error $ "failed to stream from genesis to tip: " <> show e
```

Comment:
When `MissingBlockError` takes what is now `WrongBoundError` as an argument, this can be simplified.